### PR TITLE
Add support for insert using select

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description
This PR adds support for planning and executing `INSERT` queries that take their input from a `SELECT` or `UNION` query.

```sql
insert into ... select query ...
```

## Related Issue(s)
- #9705
- Closes https://github.com/vitessio/vitess/issues/8379
- Closes https://github.com/vitessio/vitess/issues/7305
- Closes https://github.com/vitessio/vitess/issues/6330
- Closes https://github.com/vitessio/vitess/issues/5408

## Checklist
- [x] Should this PR be backported? **NO**
- [x] Tests were added or are not required 
- [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/engine/insert.go [hunk 1]
```diff
@@ -354,6 +541,66 @@ func (ins *Insert) processGenerate(vcursor VCursor, bindVars map[string]*querypb
 	return insertID, nil
 }
 
+// processGenerateFromValues generates new values using a sequence if necessary.
```
Comment (frouioui) [inline go/vt/vtgate/engine/insert.go:544]:
```suggestion
// processGenerateFromRows generates new values using a sequence if necessary.
```

File: go/vt/vtgate/engine/insert_test.go [hunk 2]
```diff
@@ -1471,3 +1488,534 @@ func TestInsertShardedUnownedReverseMapSuccess(t *testing.T) {
...
+			"sharded": {
+				Sharded: true,
+				Vindexes: map[string]*vschemapb.Vindex{
+					"hash": {
+						Type: "hash"}},
+				Tables: map[string]*vschemapb.Table{
+					"t1": {
+						ColumnVindexes: []*vschemapb.ColumnVindex{{
+							Name:    "hash",
+							Columns: []string{"id"}}}}}}}}
```
Comment (frouioui) [inline go/vt/vtgate/engine/insert_test.go:1649]: nit:
```suggestion
	invschema := &vschemapb.SrvVSchema{
		Keyspaces: map[string]*vschemapb.Keyspace{
			"sharded": {
				Sharded: true,
				Vindexes: map[string]*vschemapb.Vindex{
					"hash": {
						Type: "hash"}},
				Tables: map[string]*vschemapb.Table{
					"t1": {
						ColumnVindexes: []*vschemapb.ColumnVindex{{
							Name:    "hash",
							Columns: []string{"id"},
						}},
					},
				},
			},
		},
	}
```

Comment (frouioui) [reply go/vt/vtgate/engine/insert_test.go:1649]: here and also the other tests of the same file

File: go/vt/vtgate/planbuilder/insert.go [hunk 3]
```diff
@@ -200,12 +185,150 @@ func buildInsertShardedPlan(ins *sqlparser.Insert, table *vindexes.Table) (engin
...
+		var hasStarExpr bool
+		for _, sExpr := range sel.SelectExprs {
+			if _, hasStarExpr = sExpr.(*sqlparser.StarExpr); hasStarExpr {
+				break
+			}
+		}
+		if !hasStarExpr {
+			return vterrors.NewErrorf(vtrpcpb.Code_INVALID_ARGUMENT, vterrors.WrongValueCountOnRow, "Column count doesn't match value count at row 1")
+		}
+	}
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/insert.go:284]: shouldn't we also check that the number of expanded select expressions (from the `StarExpr` in `selectStmt`) does not go over the number of columns in `ins`?

File: go/vt/vtgate/planbuilder/insert.go [hunk 4]
```diff
@@ -232,49 +355,68 @@ func generateInsertShardedQuery(node *sqlparser.Insert, eins *engine.Insert, val
...
-		// Support the DEFAULT keyword by treating it as null
-		if _, ok := row[colNum].(*sqlparser.Default); ok {
-			row[colNum] = &sqlparser.NullVal{}
-		}
+func generateInsertSelectQuery(node *sqlparser.Insert, eins *engine.Insert) {
+	prefixBuf := sqlparser.NewTrackedBuffer(dmlFormatter)
+	suffixBuf := sqlparser.NewTrackedBuffer(dmlFormatter)
+	prefixBuf.Myprintf("insert %v%sinto %v%v ",
+		node.Comments, node.Ignore.ToString(),
+		node.Table, node.Columns)
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/insert.go:363]: Do we need a space between the comments and the ignore, or `node.Comments` already contain a trailing space?

File: go/vt/vtgate/planbuilder/testdata/onecase.txt [hunk 5]
```diff
@@ -1 +1 @@
-# Add your test case here for debugging and run go test -run=One.
+# Add your test case here for debugging and run go test -run=One.
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/testdata/onecase.txt:1]: not sure what changed here

File: go/vt/vtgate/planbuilder/testdata/dml_cases.txt [hunk 6]
```diff
@@ -3187,3 +3299,405 @@ Gen4 plan same as above
...
+      "Name": "user",
+      "Sharded": true
+    },
+    "TargetTabletType": "PRIMARY",
+    "AutoIncrement": "main:1",
+    "MultiShardAutocommit": false,
+    "TableName": "user",
+    "VindexOffsetFromSelect": {
+      "costly_map": "[-1]",
+      "name_user_map": "[-1]",
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/testdata/dml_cases.txt:3485]: I think we can exclude those `VindexOffsetFromSelect` that are unknown (with the value `-1`) from the plan output

**Review by frouioui — APPROVED** [review]
I left a bunch of comments/questions, otherwise, it looks good to me!

Thanks for migrating all the E2E tests to use the centralized `utils` package 🎉

File: go/vt/vtgate/engine/insert_test.go (hunk 2 above)
Comment (harshit-gangal) [reply go/vt/vtgate/engine/insert_test.go:1649]: This is purposely done to keep the input short for test to be read easily and referring to this without scrolling much up

File: go/vt/vtgate/planbuilder/testdata/onecase.txt (hunk 5 above)
Comment (frouioui) [reply go/vt/vtgate/planbuilder/testdata/onecase.txt:1]: I think the `\n` is missing at the end of line 1

File: go/vt/vtgate/planbuilder/insert.go (hunk 4 above)
Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/insert.go:363]: parser works fine with comments without space. if comments does not exists then the query will have 2 space.

File: go/vt/vtgate/planbuilder/insert.go (hunk 3 above)
Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/insert.go:284]: that will happen at the runtime, i.e. in the engine, it cannot happen in the planner.

**Pull request merged**